collector/extract_labels.py

```diff
@@ -0,0 +1,146 @@
+    logger.info('Number of observations: {}'.format(num_obs))
+
+    # Convert starts and durs to samples (our unit of basic time).
+    starts = [int(float(i) * sample_rate) for i in starts]
+    durs = [int(float(i) * sample_rate) for i in durs]
+
+    # Produce regexes.
+    re_discard = re.compile(discard)
+
+    # Prepend with slash if not empty string.
+    left_fmt = '\\' + left_delim if left_delim else left_delim
+    right_fmt = '\\' + right_delim if right_delim else right_delim
+    pattern = r"{}({}){}".format(left_fmt, interior, right_fmt)
+    re_delim = re.compile(pattern)
+
+    # List of all possible speakers.
+    speakers = set()
+    for text in texts:
+        for speaker in re_delim.findall(text):
+            speakers.add(speaker)
+
+    # Map speakers to integers - use OrderedDict so the output
+    # is determinant.
+    speaker_dict = OrderedDict(
+        (speaker, i)
+        for i, speaker in enumerate(
+            iter(speakers)
+        )
+    )
+
+    logger.info('Speakers found: {}'.format(speaker_dict))
+
+    # Produce result matrix.
+    labels = np.zeros((num_obs, len(speakers)), dtype=np.int32)
+
+    # Reference to the current speaker
+    current = None
+
+    # For each text, update the corresponding appropriate rows in `labels'
+    # with 1's if the speaker is speaking during them. Otherwise, put
+    # zeros, or -1 if the row is to be discarded.
+    for start_sample, dur, text in zip(starts, durs, texts):
+
+        # Find our window.
+        start_obs = start_sample // samples_per_observation
+        end_obs = (start_sample+dur) // samples_per_observation
+
+        # Current speaker may exist in this row or previous.
+        current = update_current_speaker(current, text, re_delim)
+
+        # If discard, leave -1's so we can remove rows later.
+        if not valid_text(text, re_delim, re_discard):
+            labels[start_obs:end_obs, :] = -1
+        elif current:
+            labels[start_obs:end_obs, speaker_dict[current]] = 1
+
+        logger.debug(text)
+        logger.debug('Result: {}'.format(labels[start_obs]))
+
+    logger.info('Result shape: {}'.format(labels.shape))
+    logger.info('Expected shape: ({},{})'.format(num_obs, len(speakers)))
+    logger.info('Time spent: {}'.format(time.time() - start_profile))
+
+    return labels
+
+
+def valid_text(text, re_delim, re_discard):
+    """There are 3 cases where text is not valid:
+        1. Multiple speakers observed in text.
+        2. Text doesn't start with a speaker, but has speaker in it.
+        3. The discard regex appears.
+    """
+    speakers = re_delim.findall(text)
+    on_start = re_delim.match(text)
+    discard = re_discard.search(text)
+
+    if len(speakers) > 1:
+        return False
+    if not on_start and speakers:
+        return False
+    if discard:
+        return False
+    return True
+
+
+def update_current_speaker(current, text, re_delim):
```

Comment:
Naming nitpick: `update_` implies that the function is performing some mutation; perhaps something like `get_current_speaker` is more appropriate.